Samples/MapScenarios/MapScenarios/Scenarios/WalkTime/WalkTimeViewController.swift

```diff
@@ -0,0 +1,177 @@
+            walkTimeView.removeFromSuperview()
+            return
+        }
+        
+        // Update every 5 seconds
+        DispatchQueue.main.asyncAfter(deadline: .now() + 5) { [weak self] in
+            guard let strongSelf = self else {
+                return
+            }
+            
+            strongSelf.updateWalkTimeView()
+        }
+        
+        // Set initial value
+        walkTimeView.updateWalkTime(distance: distance, averageSpeed: speedArray.average)
+    }
+    
+    func restDistance() -> CLLocationDistance {
+        // Recaculate only when the blue dot is snapping on the route path
+        guard let lastUpdateLocation = lastUpdateLocation, snappedLocation == true, let currentInstruction = self.mapView.currentRouteInstruction(), let currentIndex = self.mapView.currentRoute.routeInstructions.firstIndex(of: currentInstruction), let restInstructions = self.mapView.currentRoute.routeInstructions?[currentIndex...] else {
+            return -1
+        }
+        
+        var distance: CLLocationDistance = 0
+        
+        // The distance for rest of instructions
+        for instruction in restInstructions {
+            distance += instruction.distance
+        }
+        
+        // The distance from current location to the end of current instruction
+        if let instructionEndLocation = currentInstruction.points.last?.coordinate {
+            let userLocation = CLLocation(latitude: lastUpdateLocation.coordinate.latitude, longitude: lastUpdateLocation.coordinate.longitude);
+            let endLocation = CLLocation(latitude: instructionEndLocation.latitude, longitude: instructionEndLocation.longitude)
+            distance += userLocation.distance(from: endLocation)
+        }
+        
+        return distance
+    }
+}
+
+// MARK: - PWMapViewDelegate
+
+extension WalkTimeViewController: PWMapViewDelegate {
+    
+    func mapView(_ mapView: PWMapView!, locationManager: PWLocationManager!, didUpdateIndoorUserLocation userLocation: PWUserLocation!) {
+        lastUpdateLocation = userLocation
+    }
+    
+    func mapViewStartedSnappingLocation(toRoute mapView: PWMapView!) {
+        snappedLocation = true
+    }
+    
+    func mapViewStoppedSnappingLocation(toRoute mapView: PWMapView!) {
+        snappedLocation = false
+    }
+    
+    func mapView(_ mapView: PWMapView!, didChange instruction: PWRouteInstruction!) {
+        // Update walk time when blue dot is not acquired
+        if lastUpdateLocation == nil, let currentIndex = self.mapView.currentRoute.routeInstructions.firstIndex(of: instruction), let restInstructions = self.mapView.currentRoute.routeInstructions?[currentIndex...] {
+            var distance: Double = 0
+            for instruction in restInstructions {
+                distance += instruction.distance
+            }
+            self.walkTimeView?.updateWalkTime(distance: distance, averageSpeed: 0)
```

Comment:
why pass averageSpeed 0 here?

Reply:
This's for the case that blue dot's not acquired and end user is able to see the ETA time from the current instruction.

Reply:
Just like above, couldn't we use an average walking speed as a constant instead of zero?

Reply:
The default average speed is defined in the WalkTimeView, and `0` will be replaced with it before recalculation. The better way I think is to make the second argument optional, will do.

Reply:
@xwang-phunware will the default average speed be used if no argument is is passed to the function?

Reply:
Yes

Reply:
Good call on the optional function argument, Xiangwei!